File: frontend/src/services/inventoryService.js
import axios from '@/lib/axios';

/*
* API LẤY TỒN KHO CỦA 1 SKU CỤ THỂ
*/
export const getInventoryBySkuId = async (skuId) => {
    try {
        const response = await axios.get(`/inventory/sku/${skuId}`);
        return response.data;
    } catch (error) {
        console.error(`Lỗi lấy tồn kho của SKU ${skuId}:`, error);
        return null;
    }
};

/*
* API LẤY DANH SÁCH SKU SẮP HẾT HÀNG (CẢNH BÁO LOW STOCK)
*/
export const getLowStockItems = async () => {
    try {
        const response = await axios.get('/inventory/low-stock');
        return response.data;
    } catch (error) {
        console.error("Lỗi lấy danh sách cảnh báo tồn kho:", error);
        return { result: [] };
    }
};

/*
* API CẬP NHẬT NGƯỠNG CẢNH BÁO TỒN KHO THẤP
*/
export const updateLowStockThreshold = async (skuId, thresholdData) => {
    try {
        // thresholdData có định dạng: { threshold: 10 }
        const response = await axios.put(`/inventory/sku/${skuId}/threshold`, thresholdData);
        return response.data;
    } catch (error) {
        console.error("Lỗi cập nhật ngưỡng cảnh báo:", error);
        throw error;
    }
};

/*
* API ĐIỀU CHỈNH TỒN KHO THỦ CÔNG
*/
export const adjustStock = async (adjustData) => {
    try {
        // adjustData có định dạng: { skuId, quantityChange, reason }
        const response = await axios.post('/inventory/adjust', adjustData);
        return response.data;
    } catch (error) {
        console.error("Lỗi điều chỉnh tồn kho:", error);
        throw error;
    }
};

/*
* API LẤY LỊCH SỬ BIẾN ĐỘNG KHO CỦA 1 SKU
*/
export const getStockMovements = async (skuId) => {
    try {
        const response = await axios.get(`/inventory/sku/${skuId}/movements`);
        return response.data;
    } catch (error) {
        console.error(`Lỗi lấy biến động kho của SKU ${skuId}:`, error);
        return { result: [] };
    }
};

/*
* API LẤY BÁO CÁO TỒN KHO HIỆN TẠI (TỔNG HỢP)
*/
export const getStockOnHand = async () => {
    try {
        const response = await axios.get('/inventory/report/stock-on-hand');
        return response.data;
    } catch (error) {
        console.error("Lỗi lấy báo cáo tồn kho hiện tại:", error);
        return null;
    }
};

/*
* API LẤY BÁO CÁO ĐỊNH GIÁ KHO
*/
export const getInventoryValuation = async () => {
    try {
        const response = await axios.get('/inventory/report/valuation');
        return response.data;
    } catch (error) {
        console.error("Lỗi lấy báo cáo định giá kho:", error);
        return null;
    }
};

File: backend/src/main/java/com/example/clothingstore/service/InventoryService.java
package com.example.clothingstore.service;

import com.example.clothingstore.dtos.inventory.request.StockAdjustmentRequest;
import com.example.clothingstore.dtos.inventory.request.UpdateLowStockThresholdRequest;
import com.example.clothingstore.dtos.order.response.*;
import com.example.clothingstore.entity.Inventory;
import com.example.clothingstore.entity.Sku;

import java.util.List;

public interface InventoryService {

    /**
     * INV-001: Bootstrap bản ghi Inventory cho SKU nếu chưa có.
     * Được gọi nội bộ từ GoodsReceiptService và OrderService.
     */
    Inventory getOrCreateInventory(Sku sku);

    /**
     * INV-001: Giữ chỗ tồn kho khi đơn hàng được tạo.
     * reserved_quantity += qty; available_quantity -= qty.
     */
    void reserveStock(Long skuId, int quantity);

    /**
     * INV-001: Giải phóng tồn kho đã reserve khi đơn hàng bị hủy.
     * reserved_quantity -= qty; available_quantity += qty.
     */
    void releaseStock(Long skuId, int quantity);

    /**
     * INV-001: Xuất kho thực tế khi đơn hàng hoàn thành.
     * physical_quantity -= qty; reserved_quantity -= qty.
     */
    void deductStock(Long skuId, int quantity);

    /** INV-004: Lấy danh sách SKU có available_quantity < low_stock_threshold. */
    List<InventoryResponse> getLowStockItems();

    /** INV-004: Cập nhật ngưỡng cảnh báo tồn kho thấp cho 1 SKU. */
    InventoryResponse updateLowStockThreshold(Long skuId, UpdateLowStockThresholdRequest request);

    /** INV-005: Điều chỉnh tồn kho thủ công với audit trail. */
    StockAdjustmentResponse adjustStock(StockAdjustmentRequest request);

    /** INV-006: Xem tồn kho hiện tại của 1 SKU cụ thể. */
    InventoryResponse getInventoryBySkuId(Long skuId);

    /** INV-006: Báo cáo tồn kho tổng hợp toàn bộ SKU. */
    StockOnHandResponse getStockOnHand();

    /** INV-006: Lịch sử biến động tồn kho theo SKU. */
    List<StockMovementResponse> getStockMovements(Long skuId);

    /** INV-006: Báo cáo định giá tồn kho (available_quantity * unit_price). */
    InventoryValuationResponse getInventoryValuation();
}